### PR TITLE
Feat/typescript

### PR DESCRIPTION
**Pull request by bachstatter**
As a first step to productionise this I added typescript and converted the modules. 

Found some type errors.  Still a few unclear things but I think that can be fixed as I do more testing.

Next steps writing unit tests.

### CODE REVIEW

File: src/keeper.ts [hunk 1]
```diff
@@ -335,20 +387,20 @@ class Keeper {
     }
 
     metrics.futuresLiquidations.observe(
-      { market: this.baseAsset, success: true },
+      { market: this.baseAsset, success: "true" },
```
Comment (bachstatter) [inline src/keeper.ts:390]: Make sure this is still happy with a string

File: src/provider.ts [hunk 2]
```diff
@@ -1,17 +1,21 @@
...
+import { createLogger } from "./logging";
+
+const getErrorMessage = (e: unknown) => {
+  if (e instanceof Error) return e.message;
+  return e &&
+    typeof e === "object" &&
+    "toString" in e &&
+    typeof e.toString === "function"
+    ? e.toString()
+    : String(e);
```
Comment (bachstatter) [inline src/provider.ts:13]: is String() and toString not the same?

File: src/provider.ts [hunk 3]
```diff
@@ -113,64 +120,40 @@ class Providers {
...
             process.exit(1);
           }, HTTP_PROVIDER_TIMEOUT);
 
           // Heartbeat.
           try {
             logger.info("ping");
             stopwatch.start();
-            await runNextTick(() => provider.getBlock("latest"));
+            await runNextTick(async () => {
+              provider.getBlock("latest");
```
Comment (bachstatter) [inline src/provider.ts:147]: should await this friend

Comment (bachstatter) [reply src/provider.ts:147]: And a comment

File: src/run.ts [hunk 4]
```diff
@@ -140,27 +167,27 @@ async function run({
   metrics.trackKeeperBalance(signers[0], SynthsUSD);
 
   // Get addresses.
-  markets = markets.split(",");
+  const marketsArray = markets.split(",");
```
Comment (bachstatter) [inline src/run.ts:170]: strip the markets to be safe

File: src/signer-pool.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import { createLogger } from "./logging";
+
+class SignerPool {
+  signers: ethers.Signer[];
+  pool: number[];
+  logger: Logger;
+  constructor(signers: ethers.Signer[]) {
+    this.signers = signers;
+    this.pool = Array.from(Array(this.signers.length).keys());
+    // this.acquireCallbacks = []; unused
```
Comment (bachstatter) [inline src/signer-pool.ts:12]: Remove this friend

File: src/signer-pool.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  static async create({ signers }: { signers: ethers.Signer[] }) {
+    return new SignerPool(signers);
+  }
+
+  async acquire(): Promise<[number, ethers.Signer] | undefined> {
+    this.logger.info("awaiting signer");
+    while (!this.pool.length) {
+      await new Promise((resolve, reject) => setTimeout(resolve, 1));
+    }
+    const i = this.pool.pop();
```
Comment (bachstatter) [inline src/signer-pool.ts:25]: This should be the condition to the while loop to be really safe

File: src/signer-pool.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+  release(i: number) {
+    this.logger.info(`released signer i=${i}`);
+    this.pool.push(i);
+  }
+
+  async withSigner(cb: (signer: ethers.Signer) => Promise<void>) {
+    const [i, signer] = (await this.acquire()) || [];
+    if (!signer || i === undefined)
+      return Promise.reject("No signer available");
```
Comment (bachstatter) [inline src/signer-pool.ts:39]: Once [this](https://github.com/Synthetixio/futures-keepers/pull/7/commits/91cbfe7199d6edd4f79dac0aa91d84c88ec761bf#r773523069) comment is addressed we can remove this check

**Review by noahlitvin — APPROVED** [review]

**Pull request merged**